# Czym jest programowanie funkcyjne

W ostatnich latach została bardzo znacząco rozwinęła się nowa koncepcja programowania zwana programowaniem funkcyjnym. I dzisiejsze szkolenie pozwoli nam zapoznać sie z podstawowymi pojęciami tego co ono dotyczy.

* Okazuje się, że istnieje inny sposób formułowania myśli w kod. 
* Sposób, który w znacznej mierze rezygnuje z korzystania z pętli, warunków (w taki sposób jak my ich używamy)
* Posługujący się pojęciami takimi jak niezmienniczość, funkcja czysta, efekt uboczny.
* Wsród największych zalet jest możliwość współbierznego wykonywania naszych programów. 

## Rzeczowniki, czasowniki

W różnych opracowaniach często próbuje się budować metaforę dla tego czym jest FP dla choćby programowania obiektowego porównując rolę czasowników i rzeczowników. Uporządkujmy to

* Programowanie takie jakie znamy obecnie to programowanie imperatywne. Program składa się ze zmiennych. Na tych zmiennych za pomocą instrukcji języka wykonywane są zmiany.
* Szczególnie w programowaniu obiektowym dostrzegamy podział programu na obiekty, które się ze sobą komunikują, składają jedne z drugich i ciągle zmieniają jak długo działa nasz program.
* Te zmienne i te obiekty to rzeczowniki w tej metaforze.
* Tak jak tworzyliśmy zmienne do przechowywania danych, tak w FP tworzymy zmienne przechowujące czynności (funkcje) które chcemy wykonywać na naszych zbiorach danych.



## Stop

Rozważmy przykład i to bliski Państwa doświadczeniom. 

_W klasie odbyła się praca klasowa. I teraz mamy 20-30 prac uczniów do sprawdzenia. Jak z tym postąpimy?_

### Wersja klasyczna

Możemy w domu usiaść przy biurku. Położyć na nim stos prac, wziąć długopis i zabrać się do pracy

```python
wynik = {}
for praca in kupka_sprawdzianow:
    imie, nazwisko = praca.autor.split(' ')
    punkty = 0
    # zadanie 1
    zadanie = praca.zadanie[0]
    if zadanie.odpowiedz == 7.32:
        punkty += 1
    elif zadanie.odpowiedx == 7:
        punkty += 0.5
    ...
    # wyznacz ocene
    if punkty > 10:
        ocena = 6
    elif punkty > 8:
        ocena = 5
    ...
    wyniki[nazwisko+' '+imie] = ocena

```

## Wersja futurystyczna

Możemy rozdać uczniom prace przez nich napisane (niekoniecznie tak aby każdy dostał swoją, niekoniecznie nawet tej samej klasie) oraz przekazać im listę czynności jakie trzeba wykonać by je sprawdzić.

Zalety:
* Przyśpieszenie sprawdzania (im więcej prac tym zysk jest większy)
* Brak oszustw przy sprawdzaniu (tricky!)

Ograniczenia:
* Instrukcja musi być kompletna (gotowa przed rozpoczęciem sprawdzania).
* Po rozdaniu instrukcji nic nie można już w niej zmieniać.
* Nie można np. porównać dwóch prac.
* Nie można ocenić na dwie różne oceny tak samo rozwiązanego zadania. (patrz funkcja czysta!)
* Jeśli do sprawdzenia potrzebne są jakieś wartości to muszą być stałe (patrz niezmienniczość!)

# Wprowadzenie do programowania funkcyjnego

W dominującej części zadań systemów komputerowych wyglądają one tak, że mamy bardzo duży zbiór pewnych danych (kolekcja) i czynności jakie na nich chcemy wykonać.

W programowaniu klasycznym rozwiązalibyśmy to w sposób według schematu

```python
for element in kolekcja:
    czynnosc(element)
```

Można by podsumować opisem:
* rozkładamy duże zbiory na małe elementy, i
* wykonujemy na nich z osobna ciągi czynności

W FP unikamy rozkładania dużych kolekcji - zamiast tego korzystamy z tzw. funktorów. Schematów łączenia zbiorów danych i czynności. Przykładamy takich funktorów są:

* map - przekształć `map(czynnosc, dane)`.
* filter - wybierz `filter(sprawdz_czy_pasuje, dane)`.
* reduce - podsumuj `reduce(generuj_podsumowania, dane)`,
* inne ...

Zauważmy, że nikt nam tutaj nie gwarantuje, że np. dane będą przetwarzane w odpowiedniej kolejności. Stąd pojawia się kolejny warunek

*Czynności muszą być funkcjami czystymi* 

Powiemy, że coś jest funkcją czystą jeśli jedynym efektem przez nią wywołanym jest to co zostało przekazane przez operacje `return`

Wydaje się to dość oczywiste, ale:

* wynik funkcji musi zależeć wyłącznie od przekazanych jej parametrów,
* żadnych odwołań do (zmieniających się) zmiennych globalnych, 
* żadnych zapisów do plików, wyświetleń, odwołań do baz danych.

Wszystko powyższe i wszystko inne co narusza czystość danej funkcji nazywane jest _efektem ubocznym_. W FP kontrolujemy i monitorujemy wystąpienia efektów ubocznych.

## Lambdy

Szczególnie często w programowaniu funkcyjnym wykorzystywany jest specjalny rodzaj funkcji nazywany lambdą. To jej krótki i minimalistyczna odmiana, za to idealna jeśli chcemy 
szybko zdefiniować prostą operację.

np.


In [3]:
def czy_podzielne_przez_5(x):
    return x % 5 == 0

result = []
for liczba in (3,5,10,17):
    result.append(czy_podzielne_przez_5(liczba))
result

[False, True, True, False]

zamiast tak podanej funkcji możemy po prostu użyć lambdy

In [5]:
list(map(lambda x : x%5 == 0, (3,5,10,17)))

[False, True, True, False]

## Przykładowe użycia funktorów

Obejrzymy efekt działania kolejno map:

In [14]:
lista = [i for i in range(0,50)]
reszty = list(map(lambda i : i%6, lista))
print(reszty)

[0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1]


filter

In [15]:
lista = [i for i in range(0,50)]
podzielne_na_6 = list(filter(lambda i : i%6 == 0, lista))
print(podzielne_na_6)

[0, 6, 12, 18, 24, 30, 36, 42, 48]


i trudnego początkowo - reduce

In [16]:
from functools import reduce
suma = reduce(lambda x, y : x+ y, podzielne_na_6)
suma

216

Warto wyjaśnić jak wpływa lambda na generowanie podsumowania. Mamy początkowo tablicę 
```
[0, 6, 12, 18, 24, 30, 36, 42, 48]
```
nasz czynność jest stosowana kolejno od początku (choć nie musi być kolejno!) do elemenów naszego zbioru. Na początku pod x, y trafiają 0 oraz 6. I tablica powyżej zmienia się w 

```
[6, 12, 18, 24, 30, 36, 42, 48]
```

Po drugim użyciu
```
[18, 18, 24, 30, 36, 42, 48]
```

Po trzecim użyciu
```
[36, 24, 30, 36, 42, 48]
```
i dalej
```
[60, 30, 36, 42, 48]
[90, 36, 42, 48]
[126, 42, 48]
[168, 48]
[216]
```
Stąd wynik redukcji to 216. Bez obejrzenia sobie tego przykładu trudno sobie wyobrazić skąd bierz się nazwa reduce/redukcja.

In [11]:
list(podzielne_na_6)

[]

## Porównanie programowania imperatywnego i funkcyjnego

Rozważmy zadanie _Dla liczb od 0 do 9999 znaleźć tych których kwadraty są podzielne przez 5 lub 7 i policzyć średnią tych kwadratów._

klasyczny program miałby postać

In [17]:
suma = 0
licznik = 0
for i in range(0,9999):
    kwadrat = i*i
    if kwadrat % 7 ==0 or kwadrat % 5 == 0:
        licznik += 1
        suma += kwadrat
print(suma/licznik)

33320460.2007636


funkcyjnie wyglądałoby to tak:

In [19]:
liczby = range(0,9999)
kwadraty = map(lambda x: x*x, liczby)
kwadraty_podzielna = filter(lambda x: x%5==0 or x%7==0, kwadraty)
licznik, suma = reduce(lambda x,y : (x[0]+1, x[1]+y) , kwadraty_podzielna, (0,0) ) # (0,0) przekazane jako 3 parametr w reduce dokleja sie do tablicy od strony lewej
# fajne ćwiczenie do domu by sobie rozpisać jak to się liczy,
print(suma/licznik)

33320460.2007636


Kod funkcyjny bez wcięć 

# Comprehension list

Python posiada możliwość wyrażania kodu funkcyjnego (w pewnym zakresie), ale samo map i filter nie są za często spotykane w kodach Pythona. Jest tak dlatego, że python ma własną technikę ich zapisywania za pomocą tzw. comprehension list.

Zamiast pisać kodu powyżej - doświadczony pythonista napisałby to 


In [20]:
kwadraty = [ x*x for x in range(0,9999)] 
kwadraty_podzielna = [x for x in kwadraty if x%5==0 or x%7==0 ]
licznik, suma = reduce(lambda x,y : (x[0]+1, x[1]+y) , kwadraty_podzielna, (0,0) ) # (0,0) przekazane jako 3 parametr w reduce dokleja sie do tablicy od strony lewej
# fajne ćwiczenie do domu by sobie rozpisać jak to się liczy,
print(suma/licznik)

33320460.2007636


# Generatory

Kolejnym ciekawym elementem który pojawia się w składni Pythona, gdy mówimy o programowaniu funkcyjnym są generatory. Generatory wykorzystują instrukcje pythona wyrażoną słowem kluczowym `yield`. Jego działanie jest odrobine specyficzne. Co zobaczymy na przykładzie

In [29]:
def stworz_generator(lista):
    for element in lista:
        yield element

generator = stworz_generator([1,4,5,7,14,15])

In [30]:
print(next(generator))

1


In [31]:
print(next(generator))

4


In [32]:
print(next(generator))

5


Takiego generatora można użyć np. w pętli

In [33]:
generator = stworz_generator('Na zajęciach z CMI było dziś o programowaniu funkcyjnym. Teraz poznajemy generatory'.split(' '))

In [34]:
for element in generator:
    print(f'Mam element {element}')

Mam element Na
Mam element zajęciach
Mam element z
Mam element CMI
Mam element było
Mam element dziś
Mam element o
Mam element programowaniu
Mam element funkcyjnym.
Mam element Teraz
Mam element poznajemy
Mam element generatory


Generatory można jednak użyć do dużo ciekawszych rzeczy. Generowaniu nietrywialnych ciągów. Również tych potencjalnie nieskończonych !!

Zaprezentujemy to na przykładzie generatora rzucającego kolejnymi liczbami pierwszymi

In [39]:
def stworz_generator_liczb_pierwszych(debug=False):
    candidate = 2
    found = []
    while True:
        if debug:
            print('.', end='')
        if all(candidate % prime != 0 for prime in found): 
            yield candidate
            found.append(candidate)
        candidate += 1
liczby_pierwsze = stworz_generator_liczb_pierwszych()

In [40]:
for _, liczba in zip(range(10), liczby_pierwsze):
    print(f'Kolejna liczna pierwsza to {liczba}')

Kolejna liczna pierwsza to 2
Kolejna liczna pierwsza to 3
Kolejna liczna pierwsza to 5
Kolejna liczna pierwsza to 7
Kolejna liczna pierwsza to 11
Kolejna liczna pierwsza to 13
Kolejna liczna pierwsza to 17
Kolejna liczna pierwsza to 19
Kolejna liczna pierwsza to 23
Kolejna liczna pierwsza to 29


zauważmy ciekawą właściwość

In [42]:
liczby_pierwsze_debugowane = stworz_generator_liczb_pierwszych(debug=True)

In [43]:
next(liczby_pierwsze_debugowane)

.

2

In [44]:
next(liczby_pierwsze_debugowane)

.

3

In [45]:
next(liczby_pierwsze_debugowane)

..

5

In [46]:
next(liczby_pierwsze_debugowane)

..

7

In [47]:
next(liczby_pierwsze_debugowane)

....

11

Ktoś dostrzega co mam na myśli?

In [54]:
liczby_pierwsze = stworz_generator_liczb_pierwszych()


In [55]:
%%time
next(liczby_pierwsze)

CPU times: user 31 µs, sys: 3 µs, total: 34 µs
Wall time: 38.1 µs


2

In [56]:
for _, liczba in zip(range(5000), liczby_pierwsze):
    pass

In [57]:
%%time
next(liczby_pierwsze)

CPU times: user 10.9 ms, sys: 0 ns, total: 10.9 ms
Wall time: 10.7 ms


48623

In [58]:
%%time
next(liczby_pierwsze)

CPU times: user 10.7 ms, sys: 44 µs, total: 10.8 ms
Wall time: 10.6 ms


48647

In [59]:
%%time
next(liczby_pierwsze)

CPU times: user 9.68 ms, sys: 0 ns, total: 9.68 ms
Wall time: 9.67 ms


48649

# Przykład użycia generatorów - gra w karty

In [60]:
from random import shuffle
kolory = ['kier', 'trefl', 'karo', 'pik']
figury = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']


class Karta(object):

    def __init__(self, kolor, figura):
        self.kolor = kolor
        self.figura = figura

    def __str__(self):
        return f'{self.figura}-{self.kolor}'


def utworz_talie():
    return [Karta(kolor, figura) for figura in figury for kolor in kolory]


def tasuj(talia):
    return shuffle(talia)


class Dealer(object):

    def __init__(self, talia = None, shuffle = True):
        if talia is None:
            talia = utworz_talie()
        if shuffle:
            tasuj(talia)

        def create_generator(talia):
            while len(talia) > 0:
                yield talia.pop(0)

        self.generator = create_generator(talia)

    def nastepna(self):
        return next(self.generator)

    def split(self, no_players=2, cards='all'):
        result = [ [] for _ in range(no_players)]
        i = 0
        for card in self.generator:
            result[i % no_players].append(card)
            i += 1
            if cards != 'all' and i >= cards:
                break
        return result


d = Dealer()
gracz1, gracz2, gracz3, gracz4 = d.split(no_players=4)

In [62]:
[ str(karta) for karta in gracz1]

['5-karo',
 '2-kier',
 'A-pik',
 '5-kier',
 'J-kier',
 'Q-trefl',
 '9-trefl',
 'Q-karo',
 '8-kier',
 '7-karo',
 '3-trefl',
 'J-pik',
 '3-pik']

In [63]:
[ str(karta) for karta in gracz2]

['9-karo',
 '5-trefl',
 '7-pik',
 'J-karo',
 'K-pik',
 '4-pik',
 '6-trefl',
 '4-kier',
 '10-trefl',
 'A-trefl',
 '8-pik',
 'Q-kier',
 '9-kier']

## Przykład - gra w Black Jacka

In [64]:
from random import shuffle
kolory = ['kier', 'trefl', 'karo', 'pik']
figury = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']


class Karta(object):

    def __init__(self, kolor, figura):
        self.kolor = kolor
        self.figura = figura

    def __str__(self):
        return f'{self.figura}-{self.kolor}'

    def wartosc(self):
        if self.figura in ['2', '3', '4', '5', '6', '7', '8', '9', '10' ]:
            return [int(self.figura)]
        if self.figura in ['J', 'Q', 'K']:
            return [10]
        if self.figura == 'A':
            return [1,11]


def utworz_talie():
    return [Karta(kolor, figura) for figura in figury for kolor in kolory]


def tasuj(talia):
    return shuffle(talia)

def wartosc_kart(talia :list[Karta]):
    wartosc = [0]
    for karta in talia:
        n_wartosc = karta.wartosc()
        wartosc = [ w+v for w in n_wartosc for v in wartosc ]
    wartosci_dopuszczalne = [w for w in wartosc if w <= 21]
    warunek_przekroczenia_oczka = len(wartosci_dopuszczalne) == 0
    if warunek_przekroczenia_oczka:
        return min(wartosc)
    else:
        return max(wartosci_dopuszczalne)

class Dealer(object):

    def __init__(self, talia = None, shuffle = True):
        if talia is None:
            talia = utworz_talie()
        if shuffle:
            tasuj(talia)

        def create_generator(talia):
            while len(talia) > 0:
                yield talia.pop(0)

        self.generator = create_generator(talia)

    def nastepna(self):
        return next(self.generator)

oczko = 21
krupier_strategia_limit = 17


class GraBlackJack(object):

    def __init__(self, ilosc_talii=1):
        talia = utworz_talie()
        if ilosc_talii > 1:
            talia = [ karta for karta in talia for _ in range(ilosc_talii) ]
        self.dealer = Dealer(talia, shuffle=True)
        self.gracz = [self.dealer.nastepna() for _ in range(2)]
        self.gracz_pass = False
        self.krupier = [self.dealer.nastepna()]
        self.krupier_pass = False
        self.runda = 1

    def stan_gry(self):
        print(f'Runda {self.runda}')
        print(f'Karty gracza   mają wartosc {wartosc_kart(self.gracz)} : {[str(karta) for karta in self.gracz]}')
        print(f'Karty krupiera mają wartosc {wartosc_kart(self.krupier)} : {[str(karta) for karta in self.krupier]}')

    def koniec_gry(self):
        gracz = wartosc_kart(self.gracz)
        krupier = wartosc_kart(self.krupier)
        return (gracz > oczko or krupier > oczko) or (self.gracz_pass and self.krupier_pass)

    def wynik_gry(self):
        gracz = wartosc_kart(self.gracz)
        krupier = wartosc_kart(self.krupier)
        if gracz > oczko and krupier > oczko:
            return 'Remis'
        elif gracz > oczko:
            return 'Krupier'
        elif krupier > oczko:
            return 'Gracz'
        elif gracz == krupier:
            return 'Remis'
        elif gracz > krupier:
            return 'Gracz'
        else:
            return 'Krupier'

    def graj_runde(self):
        self.stan_gry()
        if self.koniec_gry():
            print(self.wynik_gry())
            return False
        else:
            print('<h> - hit, <s> - stand')
            ruch = input('>')
            while ruch not in ('h','s'):
                ruch = input('>')
            self.runda += 1
            if ruch == 's':
                self.gracz_pass = True
            else:
                self.gracz.append(self.dealer.nastepna())
            # gra krupiera
            if wartosc_kart(self.gracz) > oczko or self.gracz_pass:
                while wartosc_kart(self.krupier) <= oczko:
                    self.krupier.append(self.dealer.nastepna())
                    if wartosc_kart(self.krupier) >= krupier_strategia_limit:
                        self.krupier_pass = True
                        break
        return True

    def graj(self):
        while self.graj_runde():
            pass


gra = GraBlackJack()

gra.graj()

Runda 1
Karty gracza   mają wartosc 14 : ['10-pik', '4-karo']
Karty krupiera mają wartosc 5 : ['5-kier']
<h> - hit, <s> - stand
Runda 2
Karty gracza   mają wartosc 22 : ['10-pik', '4-karo', '8-karo']
Karty krupiera mają wartosc 22 : ['5-kier', '9-trefl', '8-kier']
Remis


In [65]:
gra.graj()

Runda 2
Karty gracza   mają wartosc 22 : ['10-pik', '4-karo', '8-karo']
Karty krupiera mają wartosc 22 : ['5-kier', '9-trefl', '8-kier']
Remis


In [70]:
gra = GraBlackJack()

gra.graj()

Runda 1
Karty gracza   mają wartosc 18 : ['K-pik', '8-trefl']
Karty krupiera mają wartosc 10 : ['10-trefl']
<h> - hit, <s> - stand
Runda 2
Karty gracza   mają wartosc 18 : ['K-pik', '8-trefl']
Karty krupiera mają wartosc 26 : ['10-trefl', '6-pik', 'Q-pik']
Gracz
